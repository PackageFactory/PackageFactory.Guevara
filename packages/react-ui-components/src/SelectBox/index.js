import {themr} from 'react-css-themr';
import identifiers from './../identifiers.js';
import style from './style.css';
import SelectBox from './selectBox.js';

const ThemedSelectBox = themr(identifiers.selectBox, style)(SelectBox);

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import DropDown from './../DropDown/index';
import Icon from './../Icon/index';
import IconButton from './../IconButton/index';
import TextInput from './../TextInput/index';
import SearchableSelectBox from './searchableSelectBox';
import SimpleSelectBox from './simpleSelectBox';

export default injectProps({
    DropDownComponent: DropDown,
    IconComponent: Icon,
    IconButtonComponent: IconButton,
    InputComponent: TextInput,
    SearchableSelectBoxComponent: SearchableSelectBox,
    SimpleSelectBoxComponent: SimpleSelectBox
})(ThemedSelectBox);
