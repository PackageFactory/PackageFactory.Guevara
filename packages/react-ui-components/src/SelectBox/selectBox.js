import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import DropDown from '../DropDown/index';
import DefaultSelectBoxOption from './defaultSelectBoxOption';
import mergeClassNames from 'classnames';

export default class SelectBox extends PureComponent {

    static defaultProps = {
        optionValueField: 'value',
        withoutGroupLabel: 'Without group',
        scrollable: true,
        optionComponent: DefaultSelectBoxOption
    };

    static propTypes = {
        /**
         * This prop represents a set of options.
         * Each option must have a value and can have a label and an icon.
         */
        options: PropTypes.arrayOf(
            PropTypes.shape({
                icon: PropTypes.string,
                // "value" is not part of PropTypes validation, as the "value field" is specified via the "optionValueField" property
                label: PropTypes.oneOfType([
                    PropTypes.string,
                    PropTypes.object
                ]).isRequired
            })
        ),

        /**
         * Field name specifying which field in a single "option" contains the "value"
         */
        optionValueField: PropTypes.string,

        /**
         * This prop represents the current selected value.
         */
        value: PropTypes.string,

        /**
         * This prop gets called when an option was selected. It returns the new value.
         */
        onValueChange: PropTypes.func.isRequired,

        /**
         * This prop is the placeholder text which is displayed in the selectbox when no option was selected.
         */
        placeholder: PropTypes.string,

        /**
         * This prop is an icon for the placeholder.
         */
        placeholderIcon: PropTypes.string,

        /**
         * text for the group label of options without a group
         */
        withoutGroupLabel: PropTypes.string,

        /**
         * helper for asynchronous loading; should be set to "true" as long as "options" is not yet populated.
         */
        displayLoadingIndicator: PropTypes.bool,

        /**
         * if true, allows to clear the selected element completely (without choosing another one)
         */
        allowEmpty: PropTypes.bool,

        /**
         * limit height and show scrollbars if needed, defaults to true
         */
        scrollable: PropTypes.bool,

        /**
         * search box related properties
         */
        displaySearchBox: PropTypes.bool,

        searchTerm: PropTypes.string,

        onSearchTermChange: PropTypes.func,

        /**
         * Highlight input
         */
        highlight: PropTypes.bool,

        /**
         * Component used for rendering the individual option elements; Usually this component uses "SelectBoxOption" internally for common styling.
         */
        optionComponent: PropTypes.any,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'wrapper': PropTypes.string,
            'wrapper--highlight': PropTypes.string,
            'selectedOptions': PropTypes.string,
            'selectedOptions__item': PropTypes.string
        }).isRequired, /* eslint-enable quote-props */

        //
        // Static component dependencies which are injected from the outside (index.js)
        // Used in sub-components
        //
        DropDownComponent: PropTypes.any.isRequired,
        IconComponent: PropTypes.any.isRequired,
        IconButtonComponent: PropTypes.any.isRequired,
        TextInputComponent: PropTypes.any.isRequired
    };

    constructor(...args) {
        super(...args);

        this.state = {isOpen: false};
    }

    handleDropdownToggle = e => {
        if (e.target.nodeName.toLowerCase() === 'input' && e.target.type === 'text') {
            // force dropdown open if the search-input-box is focused
            this.setState({isOpen: true});
        } else {
            this.setState({isOpen: !this.state.isOpen});
        }
    }

    handleDropdownClose = () => {
        this.setState({isOpen: false});
    }

    render() {
        const {
            options,
            value,
            optionValueField,
            displayLoadingIndicator,
            theme,
            highlight,
            placeholder,
            placeholderIcon,
            displaySearchBox,
            searchTerm,
            onSearchTermChange,
            scrollable,
            TextInputComponent,
            IconButtonComponent,
            IconComponent,
            withoutGroupLabel
        } = this.props;
        const {isOpen} = this.state;
        let allowEmpty = this.props.allowEmpty;

        const selectedValue = (options || []).find(option => option[optionValueField] === value);

        const groupedOptions = this.groupOptions(options);

        const hasMultipleGroups = Object.keys(groupedOptions).length > 1 || (Object.keys(groupedOptions).length === 1 && !groupedOptions[withoutGroupLabel]);

        // if the search box should be shown, we *need* to force allowEmpty (to display the "clear" button if a value is selected),
        // as the search box is only shown if nothing is selected.
        // If we would not force this and allowEmpty=false, the user could not go back to the search box after he has initially selected a value.
        if (displaySearchBox) {
            allowEmpty = true;
        }

        let icon = '';
        let label = '';
        if (displaySearchBox) {
            icon = 'search';
        }
        if (selectedValue) {
            label = selectedValue.label;
            icon = selectedValue.icon ? selectedValue.icon : icon;
        } else if (displayLoadingIndicator) {
            label = '[Loading]'; // TODO: localize
        } else if (placeholder) {
            label = (<span className={theme.selectBox__placeholder}>{placeholder}</span>);
            icon = placeholderIcon ? placeholderIcon : icon;
        }

        const classNames = mergeClassNames({
            [theme.wrapper]: true,
            [theme['wrapper--highlight']]: (highlight && !isOpen)
        });

        return (
            <div className={classNames}>
                <DropDown.Stateless className={theme.selectBox} isOpen={isOpen} onToggle={this.handleDropdownToggle} onClose={this.handleDropdownClose}>
                    <DropDown.Header className={theme.selectBox__btn} shouldKeepFocusState={false} showDropDownToggle={options && options.length > 0}>
                        {icon ?
                            <IconComponent className={theme.selectBox__btnIcon} icon={icon}/> :
                            null
                        }
                        {displaySearchBox && !selectedValue ?
                            <TextInputComponent
                                placeholder={placeholder}
                                value={searchTerm}
                                onChange={onSearchTermChange}
                                className={theme.selectBox__searchInput}
                                containerClassName={theme.selectBox__searchInputContainer}
                                /> :
                            <span className={theme.dropDown__itemLabel}>{label}</span>
                        }

                        {displayLoadingIndicator ?
                            <IconComponent className={theme.selectBox__loadingIcon} spin={true} icon="spinner"/> :
                            null
                        }
                        {!displayLoadingIndicator && allowEmpty && selectedValue ?
                            <IconButtonComponent className={theme.selectBox__loadingIcon} icon="times" onClick={this.handleDeleteClick}/> :
                            null
                        }
                    </DropDown.Header>
                    <DropDown.Contents className={theme.selectBox__contents} scrollable={scrollable}>
                        {hasMultipleGroups ? // skip rendering of groups if there are none or only one group
                            Object.entries(groupedOptions).map(this.renderGroup) :
                            (options || []).map(this.renderOption)}
                    </DropDown.Contents>
                </DropDown.Stateless>
            </div>
        );
    }

    /**
     * Groups the options of the selectBox by their group-attribute. Returns a javascript Map with the group names
     * as key and an array of options as values.
     * Options without a group-attribute assigned will receive the key specified in props.withoutGroupLabel.
     */
    groupOptions = options => {
        return (options || []).reduce((accumulator, currentOpt) => {
            const groupLabel = currentOpt.group ? currentOpt.group : this.props.withoutGroupLabel;
            accumulator[groupLabel] = accumulator[groupLabel] || [];
            accumulator[groupLabel].push(currentOpt);
            return accumulator;
        }, Object.create(null)); // <-- Initial value of the accumulator
    }

    /**
     * Renders the options of the selectBox as <li> and groups them below a <span>
     * that displays their group name.
     * @returns {JSX} option elements grouped by and labeled with their group-attribute.
     */
    renderGroup = group => {
        const [groupLabel, optionsList] = group;
        const {theme} = this.props;
        const groupClassName = mergeClassNames({
            [theme.selectBox__item]: true,
            [theme['selectBox__item--isGroup']]: true
        });
        return (
            <li
                key={groupLabel}
                className={groupClassName}
                >
                <span>
                    {groupLabel}
                </span>
                <ul>
                    { optionsList.map(this.renderOption) }
                </ul>
            </li>
        );
    }

    /**
     * Renders a single option (<li/>) for the select box
     * @returns {JSX} option element
     */
    renderOption = (option, index) => {
        const value = option[this.props.optionValueField];
        const {theme, IconComponent, optionComponent} = this.props;
        const onClick = () => {
            this.props.onValueChange(value);
        };

        const OptionComponent = optionComponent;
        return <OptionComponent option={option} key={index} onClick={onClick} theme={theme} IconComponent={IconComponent}/>;
    }

    handleDeleteClick = () => {
        this.props.onValueChange('');
    }
}
