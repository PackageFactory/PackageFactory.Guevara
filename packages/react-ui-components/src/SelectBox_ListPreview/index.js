/* eslint-disable camelcase, react/jsx-pascal-case */
import SelectBox_ListPreview from './selectBox_ListPreview';

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import SelectBox_CreateNew from './../SelectBox_CreateNew/index';
import SelectBox_ListPreviewFlat from './../SelectBox_ListPreviewFlat/index';
import SelectBox_ListPreviewGrouped from './../SelectBox_ListPreviewGrouped/index';

export default injectProps({
    SelectBox_CreateNew,
    SelectBox_ListPreviewFlat,
    SelectBox_ListPreviewGrouped
})(SelectBox_ListPreview);
