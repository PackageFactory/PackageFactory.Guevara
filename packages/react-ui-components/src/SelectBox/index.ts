import {themr} from '@friendsofreactjs/react-css-themr';
import keydown from 'react-keydown';

import {keys} from './config';
import identifiers from '../identifiers';
import SelectBox from './selectBox';

import style from './style.css';

const ThemedSelectBox = themr(identifiers.selectBox, style)(SelectBox);
const WithKeys = keydown(keys)(ThemedSelectBox);

export default WithKeys;
