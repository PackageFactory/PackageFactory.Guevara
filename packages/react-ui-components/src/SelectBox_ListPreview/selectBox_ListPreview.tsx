// tslint:disable:class-name
import React, {Fragment, PureComponent} from 'react';

import SelectBox_CreateNew from '../SelectBox_CreateNew';
import SelectBox_ListPreviewFlat from '../SelectBox_ListPreviewFlat';
import SelectBox_ListPreviewGrouped from '../SelectBox_ListPreviewGrouped';
import SelectBox_Option_SingleLine from '../SelectBox_Option_SingleLine';
import {SelectBoxOption, SelectBoxOptions, SelectBoxProps} from '../SelectBox/selectBox';
import {Omit} from '../../types';

export interface SelectBox_ListPreviewGroup_Props {
    readonly options: SelectBoxOptions;

    // API with SelectBox
    readonly optionValueAccessor: (option: SelectBoxOption) => string |Â undefined;
    readonly onChange: (option: SelectBoxOption) => void;
    readonly focusedValue?: string;
    readonly onOptionFocus: (option: SelectBoxOption) => void;
}

export interface SelectBox_ListPreview_Props extends Omit<SelectBoxProps, 'theme'>, SelectBox_ListPreviewGroup_Props {
    // Number of characters left to type before search
    readonly searchTermLeftToType: number;
    readonly noMatchesFound?: boolean;
    readonly searchBoxLeftToTypeLabel: string;
    readonly noMatchesFoundLabel: string;
    readonly theme?: SelectBox_ListPreview_Theme;
}

export interface SelectBox_ListPreview_Theme {
    readonly 'selectBox__item': string;
    readonly 'selectBox__item--isGroup': string;
    readonly 'selectBox__groupHeader': string;
}

/**
 * **SelectBox_ListPreview is an internal implementation detail of SelectBox**, meant to improve code quality.
 *
 * It is used inside SelectBox as the dropdown part.
 */
class SelectBox_ListPreview extends PureComponent<SelectBox_ListPreview_Props> {
    public render(): JSX.Element {
        const {
            options,
            searchTermLeftToType,
            noMatchesFound,
            noMatchesFoundLabel,
            searchBoxLeftToTypeLabel,
            theme
        } = this.props;

        const ListPreviewComponent = options.some(option => option.group !== undefined) ?
            (
                <SelectBox_ListPreviewGrouped
                    theme={theme!}
                    options={options}
                    optionValueAccessor={this.props.optionValueAccessor}
                    onChange={this.props.onChange}
                    focusedValue={this.props.focusedValue}
                    onOptionFocus={this.props.onOptionFocus}

                    withoutGroupLabel={this.props.withoutGroupLabel}
                />
            ) :
            (
                <SelectBox_ListPreviewFlat
                    theme={theme!}
                    options={options}
                    optionValueAccessor={this.props.optionValueAccessor}
                    onChange={this.props.onChange}
                    focusedValue={this.props.focusedValue}
                    onOptionFocus={this.props.onOptionFocus}
                />
            );

        // TODO: replace horible self-made I18n replace
        return (
            <Fragment>
                {searchTermLeftToType > 0 ? (
                    <div className={theme!.selectBox__item}>
                        <SelectBox_Option_SingleLine
                            option={{label: `${searchBoxLeftToTypeLabel && searchBoxLeftToTypeLabel.replace('###CHARACTERS###', `${searchTermLeftToType}`)}`, icon: 'ellipsis-h'}}
                            key={'___leftToType'}
                        />
                    </div>
                ) : ListPreviewComponent}
                {noMatchesFound && (
                    <div className={theme!.selectBox__item}>
                        <SelectBox_Option_SingleLine
                            option={{label: noMatchesFoundLabel, icon: 'ban'}}
                            key={'___noResults'}
                        />
                    </div>
                )}
                <div className={theme!.selectBox__item}>
                    <SelectBox_CreateNew
                        optionValueField={this.props.optionValueField}
                        searchTerm={this.props.searchTerm}
                        onSearchTermChange={this.props.onSearchTermChange}
                        onCreateNew={this.props.onCreateNew}
                        createNewLabel={this.props.createNewLabel}
                        focusedValue={this.props.focusedValue}
                        onOptionFocus={this.props.onOptionFocus}
                    />
                </div>
            </Fragment>
        );
    }
}

export default SelectBox_ListPreview;
