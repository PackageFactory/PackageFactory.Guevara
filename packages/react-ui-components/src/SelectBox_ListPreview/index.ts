import SelectBox_ListPreview from './selectBox_ListPreview';

export default SelectBox_ListPreview;
