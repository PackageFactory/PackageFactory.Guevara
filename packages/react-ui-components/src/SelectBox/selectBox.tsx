/* eslint-disable camelcase, react/jsx-pascal-case */
import React, {PureComponent, ReactElement} from 'react';
import {$get} from 'plow-js';
import mergeClassNames from 'classnames';
import {isEqual} from 'lodash';

import SelectBox_Option_SingleLine from '../SelectBox_Option_SingleLine';
import {PickDefaultProps} from '../../types';
import {ListPreviewElementProps} from '../ListPreviewElement/listPreviewElement';

import DropDown from '../DropDown';
import SelectBox_ListPreview from '../SelectBox_ListPreview';
import SelectBox_HeaderWithSearchInput from '../SelectBox_HeaderWithSearchInput';
import SelectBox_Header from '../SelectBox_Header';
import {SelectBox_ListPreview_Theme} from '../SelectBox_ListPreview/selectBox_ListPreview';
import {SelectBox_Header_Theme} from '../SelectBox_Header/selectBox_Header';
import {SelectBox_HeaderWithSearchInput_Theme} from '../SelectBox_HeaderWithSearchInput/selectBox_HeaderWithSearchInput';

export type SelectBoxOption = Readonly<{
    label: string | ReactElement<any>;
    icon?: string;
    disabled?: boolean;
    group?: string;
    [key: string]: any;
}>;

export type SelectBoxOptions = ReadonlyArray<SelectBoxOption>;

interface SelectBoxTheme extends SelectBox_ListPreview_Theme, SelectBox_Header_Theme, SelectBox_HeaderWithSearchInput_Theme {
    readonly 'selectBox': string;
    readonly 'selectBox--disabled': string;
    readonly 'selectBox__btn': string;
    readonly 'selectBox__btn--noRightPadding': string;
    readonly 'selectBox__contents': string;
    readonly 'selectBox__list': string;
    readonly 'wrapper': string;
    readonly 'wrapper--highlight': string;
}

export type SelectBoxOptionValueAccessor = (option: SelectBoxOption) => string;

export interface SelectBoxProps {
    // ------------------------------
    // Basic Props for core functionality
    // ------------------------------
    /**
     * This prop represents the set of options to be chosen from
     * Each option must have a value and can have a label and an icon.
     */
    readonly options: SelectBoxOptions;

    /**
     * Additional className wich will be applied
     */
    readonly className?: string;

    /**
     * Field name specifying which field in a single "option" contains the "value"
     */
    readonly optionValueField: string;

    /**
     * This prop represents the currently selected value.
     */
    readonly value?: string;

    /**
     * This prop gets called when an option was selected. It returns the new value.
     */
    readonly onValueChange: (newValue: string) => void;

    /**
     * This prop controls if the SelectBox is disabled.
     */
    readonly disabled?: boolean;

    // ------------------------------
    // Visual customization of the Select Box
    // ------------------------------

    /**
     * This prop is the placeholder text which is displayed in the selectbox when no option was selected.
     */
    readonly placeholder: string;

    /**
     * This prop is an icon for the placeholder.
     */
    readonly placeholderIcon?: string;

    /**
     * Text for the group label of options without a group
     */
    readonly withoutGroupLabel: string;

    /**
     * If true, allows to clear the selected element completely (without choosing another one)
     */
    readonly allowEmpty?: boolean;

    /**
     * Shows dropdown toggle. Set by default. Useful in components that display search, where you don't want to let the user manually controll the collapsing of selectbox
     */
    readonly showDropDownToggle?: boolean;

    /**
     * Limit height and show scrollbars if needed, defaults to true
     */
    readonly scrollable?: boolean;

    /**
     * Component used for rendering the individual option elements; Usually this component uses "ListPreviewElement" internally for common styling.
     */
    readonly ListPreviewElement?: React.ComponentClass<ListPreviewElementProps & any>; // TODO maybe enhance ListPreviewElementProps with an index signature ([key: string]: any)

    // ------------------------------
    // Asynchronous loading of data
    // ------------------------------

    /**
     * This prop is the loading text which is displayed in the selectbox when displayLoadingIndicator ist set to true.
     */
    readonly loadingLabel?: string;

    /**
     * Helper for asynchronous loading; should be set to "true" as long as "options" is not yet populated.
     */
    readonly displayLoadingIndicator: boolean;

    // ------------------------------
    // Search-As-You-Type related functionality
    // ------------------------------
    readonly displaySearchBox: boolean;
    readonly onSearchTermChange: (searchTerm: string) => void;
    readonly onSearchTermKeyPress: (event: KeyboardEvent) => void;
    readonly threshold: number;
    readonly searchTerm?: string;
    readonly searchBoxLeftToTypeLabel: string;
    readonly noMatchesFoundLabel: string;

    /**
     * Turn SelectBox into a plain input field: not showing any search results and always showing the search input. Useful in LinkEditor to be able to input links by hand.
     */
    readonly plainInputMode?: boolean;

    /**
     * If set to true, the search box is directly focussed once the SelectBox is rendered;
     * such that the user can start typing right away.
     */
    readonly setFocus?: boolean;

    // ------------------------------
    // "Create new if not exists" functionality
    // ------------------------------
    /**
     * This prop gets called when requested to create a new element
     */
    readonly onCreateNew: (value: string) => void;

    /**
     * "Create new" label
     */
    readonly createNewLabel: string;

    // ------------------------------
    // Theme & Dependencies
    // ------------------------------
    readonly theme?: SelectBoxTheme;

    /**
     * keydown // TODO comes from keydown enhancement, so maybe use decorator/enhancer
     */
    readonly keydown: {event?: KeyboardEvent};
}

type SelectBoxDefaultProps = PickDefaultProps<SelectBoxProps,
    'options' |
    'optionValueField' |
    'withoutGroupLabel' |
    'scrollable' |
    'showDropDownToggle' |
    'threshold' |
    'ListPreviewElement'
>;

export const defaultProps: SelectBoxDefaultProps = {
    options: [],
    optionValueField: 'value',
    withoutGroupLabel: 'Without group',
    scrollable: true,
    showDropDownToggle: true,
    threshold: 2,
    ListPreviewElement: SelectBox_Option_SingleLine,
};

interface SelectBoxState {
    readonly searchTerm: string;
    readonly isExpanded: boolean;
    readonly focusedValue: string;
}

const initialState: SelectBoxState = {
    searchTerm: '',
    isExpanded: false,
    focusedValue: '',
};

// TODO: document component usage && check code in detail
export default class SelectBox extends PureComponent<SelectBoxProps, SelectBoxState> {
    public static readonly defaultProps = defaultProps;

    public readonly state = initialState;

    public render(): JSX.Element {
        const {
            options,
            theme,
            showDropDownToggle,
            threshold,
            displaySearchBox,
            displayLoadingIndicator,
            ListPreviewElement,
            plainInputMode,
            disabled,
            className,
        } = this.props;

        const searchTerm = this.getSearchTerm();

        const {focusedValue} = this.state;
        const isExpanded = disabled ? false : this.state.isExpanded;

        const headerClassName = mergeClassNames(
            className,
            theme!.selectBox__btn,
            {
                [theme!['selectBox__btn--noRightPadding']]: !showDropDownToggle,
                [theme!['selectBox--disabled']]: disabled
            }
        );

        const optionValueAccessor = this.getOptionValueAccessor();

        const searchTermLeftToType = displaySearchBox ? threshold - searchTerm.length : 0;
        const noMatchesFound = searchTermLeftToType > 0 || displayLoadingIndicator ? false : !options.length;

        return (
            <DropDown.Stateless className={theme!.selectBox} isOpen={isExpanded} onToggle={this.handleToggleExpanded} onClose={this.handleClose}>
                <DropDown.Header className={headerClassName} shouldKeepFocusState={false} showDropDownToggle={showDropDownToggle && options.length > 0}>
                    {this.renderHeader()}
                </DropDown.Header>
                <DropDown.Contents className={theme!.selectBox__contents} scrollable={true}>
                    {!plainInputMode &&
                        <ul className={theme!.selectBox__list}>
                            <SelectBox_ListPreview
                                {...this.props}

                                theme={theme}
                                optionValueAccessor={optionValueAccessor}
                                ListPreviewElement={ListPreviewElement}
                                focusedValue={focusedValue}
                                onChange={this.handleChange}
                                onOptionFocus={this.handleOptionFocusChange}
                                searchTermLeftToType={searchTermLeftToType}
                                noMatchesFound={noMatchesFound}
                                searchTerm={searchTerm}
                            />
                        </ul>}
                </DropDown.Contents>
            </DropDown.Stateless>
        );
    }

    private getSearchTerm(): string {
        return this.props.searchTerm || this.state.searchTerm;
    }

    private getOptionValueAccessor(): SelectBoxOptionValueAccessor {
        // returns a function that extract a value from the object
        // TODO extend $get path to return a curried function if called with _path_ arg only
        // @ts-ignore
        return $get([this.props.optionValueField]);
    }

    private renderHeader(): JSX.Element {
        const {
            displaySearchBox,
            displayLoadingIndicator,
            options,
            value,
            allowEmpty,
            plainInputMode,
            disabled,
        } = this.props;
        const searchTerm = this.getSearchTerm();
        const optionValueAccessor = this.getOptionValueAccessor();

        // Compare selected value less strictly: allow loose comparision and deep equality of objects
        // tslint:disable-next-line:triple-equals
        const selectedOption = options.find(option => optionValueAccessor(option) == value || isEqual(optionValueAccessor(option), value));

        if (displaySearchBox && (!value || plainInputMode)) {
            return (
                <SelectBox_HeaderWithSearchInput
                    {...this.props}
                    disabled={disabled}
                    onSearchTermChange={this.handleSearchTermChange}
                    searchTerm={searchTerm}
                    onKeyDown={this.handleKeyDown}
                />
            );
        }

        const showResetButton = Boolean(allowEmpty && !displayLoadingIndicator && value);

        return (
            <SelectBox_Header
                {...this.props}
                option={selectedOption}
                showResetButton={showResetButton}
                onReset={this.handleDeleteClick}
            />
        );
    }

    private readonly handleChange = (option: SelectBoxOption) => {
        const optionValueAccessor = this.getOptionValueAccessor();
        this.props.onValueChange(optionValueAccessor(option));
        this.setState({
            searchTerm: ''
        });
    }

    private readonly handleDeleteClick = (event?: MouseEvent) => {
        if (event) {
            // Don't open SelectBox on value clear
            event.stopPropagation();
        }
        this.props.onValueChange('');
    }

    private readonly handleToggleExpanded = () => {
        // Return early if disabled
        if (this.props.disabled) {
            return;
        }

        if (this.props.displaySearchBox) {
            if (this.props.value) {
                // When click on header in search mode with value selected, clear it
                this.props.onValueChange('');
                this.setState({
                    isExpanded: true,
                });
            } else {
                // Force expanded dropdown unless has showDropDownToggle (e.g. for nodetypes filter in the PageTree)
                this.setState({
                    isExpanded: this.props.showDropDownToggle ? !this.state.isExpanded : true,
                });
            }
        } else {
            // If simple SelectBox, just toggle it
            this.setState({
                isExpanded: !this.state.isExpanded,
            });
        }
    }

    private readonly handleClose = () => {
        this.setState({
            isExpanded: false
        });
    }

    private readonly handleOptionFocusChange = (option: SelectBoxOption) => {
        const optionValueAccessor = this.getOptionValueAccessor();
        this.setState({
            focusedValue: optionValueAccessor(option)
        });
    }

    private readonly handleSearchTermChange = (searchTerm: string) => {
        if (searchTerm.length >= this.props.threshold) {
            this.props.onSearchTermChange(searchTerm);
        } else {
            this.props.onSearchTermChange('');
        }
        this.setState({
            isExpanded: Boolean(searchTerm),
            searchTerm
        });
        this.props.onSearchTermChange(searchTerm);
    }

    /**
     * Keyboard handling
     */
    public componentWillReceiveProps({keydown}: SelectBoxProps): void {
        if (keydown.event) {
            this.handleKeyDown(keydown.event);
        }
    }

    private readonly handleKeyDown = (e: KeyboardEvent) => {
        const {options, onSearchTermKeyPress} = this.props;
        if (typeof onSearchTermKeyPress === 'function') {
            // Pass through keydown event, needed for keyboard handling
            onSearchTermKeyPress(e);
        }
        if (this.state.isExpanded && e && ['ArrowDown', 'ArrowUp', 'Enter', 'Escape'].includes(e.key)) {
            // Do not scroll while we are doing keyboard interaction
            e.preventDefault();

            const optionValueAccessor = this.getOptionValueAccessor();
            const currentIndex = options.findIndex(option => optionValueAccessor(option) === this.state.focusedValue);

            if (e.key === 'ArrowDown') {
                const newIndex = currentIndex + 1 >= options.length ? currentIndex : currentIndex + 1;
                this.setState({
                    focusedValue: optionValueAccessor(options[newIndex])
                });
            } else if (e.key === 'ArrowUp') {
                const newIndex = currentIndex - 1 < 0 ? 0 : currentIndex - 1;
                this.setState({
                    focusedValue: optionValueAccessor(options[newIndex])
                });
            } else if (e.key === 'Enter') {
                if (currentIndex < options.length && currentIndex >= 0) {
                    this.handleChange(options[currentIndex]);
                }

                this.setState({
                    isExpanded: false
                });
            } else if (e.key === 'Escape') {
                this.setState({
                    focusedValue: '',
                    isExpanded: false
                });
            }
        }
    }
}
