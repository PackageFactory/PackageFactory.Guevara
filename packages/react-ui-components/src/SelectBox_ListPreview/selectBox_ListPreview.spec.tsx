import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import SelectBox_ListPreview, {SelectBox_ListPreview_Props} from './selectBox_ListPreview';
import {SelectBoxOptions, SelectBoxOption} from '../SelectBox/selectBox';

describe('SelectBox_ListPreview', () => {
    const props: SelectBox_ListPreview_Props = {
        options: [],
        onChange: jest.fn(),
        searchTermLeftToType: 2,
        searchBoxLeftToTypeLabel: 'left to type',
        noMatchesFoundLabel: 'no matches',
        theme: {
            'selectBox__item': 'selectBox__item-ThemeClassName',
            'selectBox__item--isGroup': 'selectBox__item--isGroup-ThemeClassName',
            'selectBox__groupHeader': 'selectBox__groupHeader-ThemeClassName',
        },
        withoutGroupLabel: 'ungrouped',
        optionValueField: 'testValue',
        onSearchTermChange: jest.fn(),
        onCreateNew: jest.fn(),
        createNewLabel: 'create new',
        onOptionFocus: jest.fn(),
        displayLoadingIndicator: false,
        displaySearchBox: false,
        keydown: {},
        onSearchTermKeyPress: jest.fn(),
        onValueChange: jest.fn(),
        optionValueAccessor: jest.fn(),
        placeholder: 'placeholder',
        threshold: 2,
    };

    it('should render correctly empty.', () => {
        const wrapper = shallow(<SelectBox_ListPreview {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly with grouped options.', () => {
        const options: SelectBoxOptions = [
            {
                label: 'option1',
                testValue: 'value1',
                group: 'grouped',
            },
            {
                label: 'option2',
                testValue: 'value2',
                group: 'grouped',
                icon: 'level-up',
            },
            {
                label: 'option3',
                testValue: 'value3',
                disabled: true,
            },
        ];

        const wrapper = shallow(<SelectBox_ListPreview {...props} options={options} />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly with ungrouped options.', () => {
        const options: SelectBoxOptions = [
            {
                label: 'option1',
                testValue: 'value1',
            },
            {
                label: 'option2',
                testValue: 'value2',
                icon: 'level-up',
            },
            {
                label: 'option3',
                testValue: 'value3',
                disabled: true,
            },
        ];
        const valueFied = 'testValue';
        const optionValueAccessor = (option: SelectBoxOption) => option[valueFied];

        const wrapper = shallow(<SelectBox_ListPreview {...props} options={options} optionValueAccessor={optionValueAccessor}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
