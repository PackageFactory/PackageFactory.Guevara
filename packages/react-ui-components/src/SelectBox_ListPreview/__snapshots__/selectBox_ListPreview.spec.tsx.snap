// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox_ListPreview should render correctly empty. 1`] = `
<Fragment>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_Option_SingleLine
      key="___leftToType"
      option={
        Object {
          "icon": "ellipsis-h",
          "label": "left to type",
        }
      }
    />
  </div>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_CreateNew
      createNewLabel="create new"
      onCreateNew={[MockFunction]}
      onOptionFocus={[MockFunction]}
      onSearchTermChange={[MockFunction]}
      optionValueField="testValue"
    />
  </div>
</Fragment>
`;

exports[`SelectBox_ListPreview should render correctly with grouped options. 1`] = `
<Fragment>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_Option_SingleLine
      key="___leftToType"
      option={
        Object {
          "icon": "ellipsis-h",
          "label": "left to type",
        }
      }
    />
  </div>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_CreateNew
      createNewLabel="create new"
      onCreateNew={[MockFunction]}
      onOptionFocus={[MockFunction]}
      onSearchTermChange={[MockFunction]}
      optionValueField="testValue"
    />
  </div>
</Fragment>
`;

exports[`SelectBox_ListPreview should render correctly with ungrouped options. 1`] = `
<Fragment>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_Option_SingleLine
      key="___leftToType"
      option={
        Object {
          "icon": "ellipsis-h",
          "label": "left to type",
        }
      }
    />
  </div>
  <div
    className="selectBox__item-ThemeClassName"
  >
    <SelectBox_CreateNew
      createNewLabel="create new"
      onCreateNew={[MockFunction]}
      onOptionFocus={[MockFunction]}
      onSearchTermChange={[MockFunction]}
      optionValueField="testValue"
    />
  </div>
</Fragment>
`;
